File: Gui/src/main/java/org/workcraft/gui/propertyeditor/choice/ChoiceProperty.java
package org.workcraft.gui.propertyeditor.choice;

import java.awt.Component;

import org.workcraft.dependencymanager.advanced.user.ModifiableExpression;
import org.workcraft.gui.propertyeditor.EditableProperty;
import org.workcraft.gui.propertyeditor.GenericCellEditor;
import org.workcraft.gui.propertyeditor.GenericEditorProvider;
import org.workcraft.gui.propertyeditor.RendererProvider;
import org.workcraft.gui.propertyeditor.string.StringProperty;
import org.workcraft.util.Action;
import org.workcraft.util.Pair;

import pcollections.PVector;

public class ChoiceProperty {
	private static final <T> GenericEditorProvider<T> createEditorProvider(final PVector<Pair<String,T>> choice) { 
		return new GenericEditorProvider<T>(){

			@Override
			public GenericCellEditor<T> createEditor(T initialValue, Action accept, Action cancel) {
				return new ChoiceCellEditor<T>(initialValue, choice, accept);
			}
		};
	}
	
	private static final <T> RendererProvider<T> createRendererProvider(final PVector<Pair<String,T>> choice) {
		return new RendererProvider<T>() {
			String lookupValue(PVector<Pair<String,T>> choice, T value) {
				for(Pair<String,T> p : choice)
					if(p.getSecond().equals(value))
						return p.getFirst();
				throw new RuntimeException("The property value was outside of the specified choice");
			}
			
			@Override
			public Component createRenderer(T value) {
				return StringProperty.RENDERER_PROVIDER.createRenderer(lookupValue(choice, value));
			}
		};
	}

	public static <T> EditableProperty create(String name, PVector<Pair<String,T>> choice, ModifiableExpression<T> property) {
		return EditableProperty.Util.create(name, createEditorProvider(choice), createRendererProvider(choice), property);
	}
}

File: Gui/src/main/scala/org/workcraft/gui/propertyeditor/choice/ChoiceProperty.scala
package org.workcraft.gui.propertyeditor.choice

import java.awt.Component
import org.workcraft.dependencymanager.advanced.user.ModifiableExpression
import org.workcraft.gui.propertyeditor.EditableProperty
import org.workcraft.gui.propertyeditor.GenericCellEditor
import org.workcraft.gui.propertyeditor.GenericEditorProvider
import org.workcraft.gui.propertyeditor.RendererProvider
import org.workcraft.gui.propertyeditor.string.StringProperty
import org.workcraft.util.Action
import org.workcraft.util.Pair
import pcollections.PVector
import org.workcraft.scala.effects.IO
import org.workcraft.scala.Expressions.Expression
import org.workcraft.scala.Expressions._

object ChoiceProperty {

  def apply[T](name: String, choice: List[(String, T)], property: ModifiableExpression[T]): Expression[EditableProperty] = property.map(value => {

    val e = new GenericEditorProvider[T] {
      override def createEditor(initialValue: T, accept: IO[Unit], cancel: IO[Unit]): GenericCellEditor[T] =
        new ChoiceCellEditor[T](initialValue, choice, accept)
    }

    val r = new RendererProvider[T] {
      val m = choice.map(_.swap).toMap
      def createRenderer(value: T): Component = StringProperty.RendererProvider.createRenderer(m(value))
    }

    EditableProperty(name, e, r, value, property.set(_: T))
  })
}
